프로젝트 관리 앱 [STEP 2-1] Judy

Judy-999 opened the pull request:
안녕하세요 라자냐(@wonhee009) 추석 잘 보내셨나요? 🤗
프로젝트 관리 앱 첫 PR 제출합니다!
처음 적용해보는 것이 많아 만족스럽지 못해 계속 수정해보고 싶었지만
중간에 라자냐의 피드백을 받고 수정하는게 좋을 것 같아 제출합니다 🥲
<br>

## 코드 설명 👩🏻‍💻
**Model**
- `Work` : 해야 할 일을 나타내는 모델 타입 [title, content, deadline]
- `SampleData` : UI 구현을 확인하기 위한 샘플 데이터를 담은 타입
<br>

**View**
- `ProjectManagerView` : [TODO, DOING, DONE]의 테이블 뷰를 가진 메인 화면에 보여지는 뷰
- `HeaderView` : [TODO, DOING, DONE]의 타이틀과 현재 셀 개수를 나타내는 뷰
- `WorkTableViewCell` : 할 일의 내용을 표현하는 테이블 뷰 셀
- `WorkManageView` : 할 일의 세부정보를 표현하거나, 새로운 할 일을 추가하는 뷰
<br>

**ViewModel**
- `WorkViewModel` : 상태별 할 일의 데이터를 가지고 변경을 담당
<br>

**Extension**
- `Date` 
	- `convertToRegion` : 지역에 따른 날짜 형식으로 변경하는 메서드
	- `checkOverdue` : 오늘 날짜와 비교해 기한이 지났는지 판단하는 메서드
- `UIView`
	- `applyShadow` : 테두리에 그림자 효과를 넣어주는 메서드
<br>

## 구현 화면 📱
**메인 화면**
<img src = "https://i.imgur.com/GpXZ0iz.png" width="480" height="350">

**할 일 상세보기 및 생성**
![](https://i.imgur.com/Ke9bPZc.gif)

**지역별 날짜 표현**


| US로 설정한 메인화면 | US로 설정한 상세화면 | 
| -------- | -------- |
|  ![](https://i.imgur.com/nMvbIaa.png)|  ![](https://i.imgur.com/YT0Og7E.png) |

<br>

## 고민한 내용 🤔
### 1. 기한 지남 여부
오늘을 기준으로 기한이 지난 할 일의 날짜는 빨간색으로 표시해야 해서 `Date`를 extension하여 기한 지남 여부를 `Bool` 값으로 반환하는 `checkOverdue` 메서드를 구현했습니다. 처음에는 단순히 `Date()`와 대소비교를 하였는데, 오늘 날짜여도 시간이 지났으면 기한이 지났음으로 표시되는 문제가 있었습니다. 시간에 대한 기한 설정은 없으므로 날짜로만 비교하는게 맞다고 생각해 대소비교와 함께 오늘 날짜인지 비교하는 로직을 추가했습니다.
<br>

```swift
func checkOverdue() -> Bool {
    let today = Date().convertToRegion()
    let dateToCompare = self.convertToRegion()
        
    return today != dateToCompare && Date() >= self
}
```
<br>

시간을 제외하고 날짜가 오늘과 같은지 비교하는 다양한 방법이 있지만 간단하게 미리 구현했던 `convertToRegion` 메서드로 `String`으로 변경된 날짜가 같은지 비교하는 방식으로 했습니다.
<br>

### 2. HeaderView
![](https://i.imgur.com/0khGaAW.png)

할 일의 상태 제목과 개수를 나타내는 부분은 커스텀뷰로 만들어 사용했습니다.
개수를 표현하는 이미지는 고민 후에 SFSymbol에서 `0.circle.fill`으로 된 이미지를 사용하여 개수에 따라 숫자를 변경하여 사용하기로 했습니다. 
<br>


## 궁금한 내용 💭
RxSwift와 MVVM 구조를 처음 적용해보면서 코드 자체보다 설계에서 어려움이 있었습니다. 저의 공부 부족으로 인해 해결되지 않은 궁금점일 수도 있지만 라자냐에게 질문을 통해 이해해보고자 합니다..!

### <br>1. View-ViewController 사이 접근제어자
MVVM 구조에서 `ViewController`는 뷰로써 역할을 한다고 알고 있습니다. 이번 프로젝트에서 스토리보드 없이 코드로 뷰를 구현하기로 하여 `UIView`를 상속받은, 이전의 MVC 구조에서 `View` 역할을 하던 코드도 있습니다.

`ViewController`에서 뷰모델을 가지고 있고, 뷰컨트롤러의 `view`는 `ProjectManagerView`와 같은 커스텀 뷰로 대체하여 사용하고 있습니다. 그렇다보니 뷰컨트롤러에서 `ProjectManagerView`의 프로퍼티에 bind 작업을 하려면 직접 접근을 해야 해서 `private` 접근 제어자로 보호할 수가 없습니다.
<br>

```swift
// ProjectManagerView.swift
final class ProjectManagerView: UIView {
    // MARK: - Properties
    let toDoTitleView = HeaderView()
    let doingTitleView = HeaderView()
    let doneTitleView = HeaderView()
	...
}
```

```swift
// ProjectManagerViewController.swift
viewModel.todoCountImage
    .observe(on: MainScheduler.instance)
    .bind(to: projectManagerView.toDoTitleView.countImageView.rx.image)
    .disposed(by: disposeBag)
```
<br>

뷰컨트롤러 역시 뷰의 역할을 하기 때문에 뷰끼리는 직접 접근해도 문제될 게 없지 않을까 싶으면서도 프로퍼티는 최대한 `private`로 숨기라는 리뷰를 받은 적도 있어서 고민이 됩니다..!
<br>

### 2. Bind의 방식 및 위치 
위와 같은 맥락일 수 있는 질문으로 UI 요소에 bind 하는 작업은 `-View`와 `-ViewController` 중 어느 곳에서 실행하는 게 좋을지 궁금합니다.
<br>

```swift
// WorkManageView.swift
let work = PublishSubject<Work>()

private func bind() {
    work.observe(on: MainScheduler.instance)
        .subscribe(onNext: { [weak self] work in
            guard let self = self else { return }
            self.titleTextField.text = work.title
            self.contentTextView.text = work.content
            self.deadlinePicker.date = work.deadline
        })
        .disposed(by: disposeBag)
}
```
```swift
// ManageWorkViewController.swift
func configureWork(_ work: Work) {
    workManageView.work.onNext(work)
}
```
<br>

실제로 할 일을 선택했을 시 할 일의 상세내용을 뷰에 표현할 때 `WorkManageView`가 `PublishSubject<Work>`를 가지고 뷰컨트롤러에서 `onNext(work)`로 넘겨주기만 했습니다. 

이와 같이 해당 UI 요소를 포함하는 뷰에서 bind(혹은 subscribe) 작업을 하고 뷰컨에서는 데이터만 넘겨주는게 좋을지 궁금합니다.

만약 각각 뷰에서 구현한다면 위처럼 Observable을 프로퍼티로 가지는 방식과 뷰모델을 직접 넘겨주는 방식  중에서 어떤 방법이 많이 사용될까요? 혹은 이외에도 다른 좋은 방법이 있을까요?

뷰와 뷰컨트롤러의 역할 분담에 대해서 어려움을 느끼고 있습니다 🥲
<br>

### 3. 중복되는 코드 간소화
TODO, DOING, DONE이라는 할 일의 상태에 따라 테이블뷰를 따로 가지고 있습니다. 

그렇다보니 아래 코드와 같이 같은 작업을 세 번씩 반복하고 있습니다 😅
<br>

```swift
viewModel.todoCountImage
    .asDriver(onErrorJustReturn: nil)
    .drive(projectManagerView.toDoTitleView.countImageView.rx.image)
    .disposed(by: disposeBag)

viewModel.doingCountImage
    .asDriver(onErrorJustReturn: nil)
    .drive(projectManagerView.doingTitleView.countImageView.rx.image)
    .disposed(by: disposeBag)

viewModel.doneCountImage
    .asDriver(onErrorJustReturn: nil)
    .drive(projectManagerView.doneTitleView.countImageView.rx.image)
    .disposed(by: disposeBag)
```
<br>

동일한 코드가 반복되는 느낌인데 바라보는 데이터와 UI 요소가 다르기 때문에 어쩔 수 없는건지, 간소화할 수 있는 방법이 있을지 궁금합니다!

---

제가 개념을 제대로 이해하지 못하여 질문 자체가 이상할 수 있는데 라자냐의 생각이나 공부해봐야 할 부분을 알려주시면 감사하겠습니다 🙇‍♀️

Judy-999 committed:
chore: 빌드 경고 제거

- 'IPHONES_DEVELOPMENT_TARGET' 경고 해결
- Pod 프로젝트의 경고 표시 무시

Judy-999 committed:
chore: 스토리보드 및 주석 제거

- 네비게이션 컨트롤러를 첫 화면으로 설정
- AppDelegate와 SceneDelegate의 주석 제거

Judy-999 committed:
feat: 테이블뷰 위에 넣을 HeaderView 생성

- 할 일의 분류와 개수 표시

Judy-999 committed:
feat: 메인 화면인 ProjectManagerView 생성

Judy-999 committed:
refactor: HeaderView의 constraint 변경

- horizontalStackView를 사용하지 않고 제약을 설정

Judy-999 committed:
refactor: navigationbar와 view가 겹치는 문제 수정

- ProjectManagerView의 constraint 변경
- ViewController의 이름을 ProjectManagerView로 변경

Judy-999 committed:
feat: Work 모뎉 및 WorkTableViewCell 생성

Judy-999 committed:
feat: UI 확인을 위한 샘플 데이터(SampleData) 추가

Judy-999 committed:
feat: 뷰모델 생성 및 테이블뷰의 내용 채우기

- ProjectManagerView의 tableView 접근제어자 및 constraint 변경

Judy-999 committed:
feat: HeaderView의 숫자 이미지가 셀 개수에 따라 변경되도록 수정

Judy-999 committed:
feat: 할 일을 추가 및 추가하는 WorkManageView 생성

- ManageWorkViewContoroller : Cancel, Done 바버튼 추가

Judy-999 committed:
feat: 메인 화면에 할 일을 추가하는 '+' 바버튼 생성

Judy-999 committed:
feat: 할 일 터치 시 수정 화면에 내용을 채우고 띄우는 코드 추가

Judy-999 committed:
feat: Date를 extension하여 지역 형식에 따라 표시되도록 변경

Judy-999 committed:
feat: UIView를 extension하여 그림자 넣는 메서드 추가

Judy-999 committed:
feat: 기한을 지난 날짜는 빨간색으로 표시되도록 변경

- 오늘이 기한인 경우는 기한이 지나지 않은 것으로 설정

Judy-999 committed:
refator: 테이블 별 헤더 이미지를 뷰모델에서 변경하도록 수정

Judy-999 committed:
refactor: ProjectManagerViewController의 메서드 분리

- setupBinding 메서드를 UI 별로 메서드 분리
- ManageWorkViewController를 띄우는 코드를 메서드로 분리

Judy-999 committed:
feat: WorkTableViewCell에 prepareForReuse 구현

- DisposeBag을 갱신

Judy-999 committed:
chore: line_length 제한 준수

Judy-999 committed:
refactor: 클래스에 final 키워드 추가 및 세 줄 제한

- 레이블의 numberOfLines을 3으로 하여 내용이 세 줄이 넘어가면 ...으로 표시되도록 변경

Judy-999 committed:
refactor: UIImage의 강제언래핑 제거

Judy-999 committed:
refactor: count 이미지에 drive 방식 적용

Judy-999 committed:
docs: Step 2-1 README 업데이트

wonhee009 commented on ProjectManager/ProjectManager/Extension/Date+Extension.swift:
```diff
@@ -0,0 +1,25 @@
+        let dateToCompare = self.convertToRegion()
+        
+        return today != dateToCompare && Date() >= self
+    }
```
해당 메서드가 `Date` extension으로 구현되어 있어 모든 `Date` type에서 해당 메서드를 사용할 수 있을거 같습니다.
`checkOverdue()` 메서드가 모든 `Date` 타입에서 사용되야할 메서드인가요?

Judy-999 replied:
`checkOverdue` 오늘과 비교하여 기한을 지났는지 판단하는 메서드라 굳이 모든 `Date` 타입에서 사용될 메서드는 아닌 것 같네요..!

`DateManager`라는 타입을 만들어 메서드를 이동시켰습니다.

22a1987b3ce546f54b0e8f6000faa1421800e780

wonhee009 commented on ProjectManager/ProjectManager/View/HeaderView.swift:
```diff
@@ -0,0 +1,69 @@
+        setupView()
+    }
+    
+    required init?(coder: NSCoder) {
+        super.init(coder: coder)
+        setupView()
+    }
```
프로젝트 설명에는 모든 view를 코드로 작성했다고 본거 같은데요.
해당 init에서도 `init(frame: CGRect)`와 동일하게 뷰 초기세팅을 해주고 있는데요.
두 init은 어떤 차이점이 있나요?

Judy-999 replied:
제가 알고 있는 차이점은 다음과 같습니다!

1. `init(frame: CGRect)`
코드로 뷰를 만드는 경우 `init(frame: CGRect)`를 통해 초기화됩니다. Interface Builder를 통해 디자인한 경우에는 이 메서드는 호출되지 않습니다.

2. `init?(coder: NSCoder)`
스토리보드나 nib 파일을 사용해 뷰를 커스텀 했을 때는 해당 init으로 초기화됩니다. NSCoding을 채택하는 경우 모두  `init?(coder: NSCoder)`를 구현해줘야 하는데, UIView가 NSCoding을 채택하고 있기 때문에 UIView를 상속받은 클래스에서도 구현이 필요합니다.

이 프로젝트에선 코드로 뷰를 구현했기 때문에 `init?(coder: NSCoder)`은 호출되지 않겠지만 이전 리뷰에서 호출되지 않는다고 작동하지 않는 메서드로 비워두는 것보단 내용을 채워놓는 것도 좋다고 하셔서 초기세팅을 같이 해주었습니다!

wonhee009 commented on ProjectManager/ProjectManager/View/HeaderView.swift:
```diff
@@ -0,0 +1,69 @@
+        addSubView()
+        setupConstraints()
+        self.backgroundColor = .systemBackground
+    }
```
메서드 호출 순서가 `setupView` -> `addSubView` -> `setupConstraints`인것으로 이해했는데요.
메서드들 작성 위차가 호출 순서와 상이하여 헷갈리는거 같습니다.
코드가 어떤 흐름으로 작성되어 있는가도 가독성을 올리는데 중요한 요소라고 생각합니다 :)

Judy-999 replied:
`addSubView `와 `setupConstraints`를 구현한 후 둘을 한 번에 처리하는 `setupView`를 구현해서 위와 같은 순서가 되었습니다.

의미상도 가독성 측면에서도 `setupView` -> `addSubView` -> `setupConstraints` 순서가 맞아 변경했습니다.

54e1372714d23cba99baa2d842ec913dd0cf03c7

wonhee009 commented on ProjectManager/ProjectManager/View/ProjectManagerView.swift:
```diff
@@ -0,0 +1,126 @@
+        stackView.alignment = .fill
+        stackView.spacing = 8
+        return stackView
+    }()
```
`HeaderView + TableView`를 넣는 StackView로 이해했는데요.
StackView 사용하지 않고 TableView에 headerView 지정해주는 방식으로는 UI 구현이 어렵나요?.?

Judy-999 replied:
커스텀한 뷰를 사용하면서 뷰와 뷰를 스택뷰로 묶는다는 생각만 하고 헤더에 뷰를 넣을 생각을 못했네요!!

headerView로 지정하여 스택뷰 없이 뷰를 구성할 수 있도록 변경했습니다!

wonhee009 commented on ProjectManager/ProjectManager/View/ProjectManagerView.swift:
```diff
@@ -0,0 +1,126 @@
+
+import UIKit
+
+final class ProjectManagerView: UIView {
```
3개의 HeaderView, 3개의 TableView를 생성 후, StackView에 넣어주고 기본적인 설정 (배경색, static한 타이틀 설정)을 하고 있는거로 이해했습니다.
이런 역할을 하는 View를 생성한 이유가 있을까요?
저는 custom한 HeaderView를 별개의 파일로 생성하는 이유는 재사용성을 위해서라고 생각하는데요.
`ProjectManagerView`는 프로젝트 관리 화면만을 위한 view라는 생각이 들고 이 부분은 `ProjectManagerViewController`에서 충분히 조합해서 사용할 수 있는 부분이라는 생각이 들어서 질문 드립니다 :)

Judy-999 replied:
뷰는 뷰대로, 뷰컨은 뷰컨대로 역할을 나눈다는게 뷰컨에서 뷰를 조합해서 사용한다는 생각을 못한 것 같습니다..!

`ProjectManagerView`를 삭제하고 테이블 뷰를 커스텀한 `WorkTableView`를 구현해 뷰컨에서 `HeaderView`와 함께 생성하여 뷰를 만들어 사용하는 방식으로 변경했습니다.

wonhee009 commented on ProjectManager/ProjectManager/ViewModel/WorkViewModel.swift:
```diff
@@ -0,0 +1,35 @@
+    let doingWorks = BehaviorRelay<[Work]>(value: [])
+    let doneWorks = BehaviorRelay<[Work]>(value: [])
+    
+    lazy var todoCountImage = todoWorks.map {
+        UIImage(systemName: "\($0.count).circle.fill")
+    }
+    
+    lazy var doingCountImage = doingWorks.map {
+        UIImage(systemName: "\($0.count).circle.fill")
+    }
+
+    lazy var doneCountImage = doneWorks.map {
+        UIImage(systemName: "\($0.count).circle.fill")
+    }
```
`lazy`로 정의한 이유가 있을까요?

Judy-999 replied:
`todoWorks`, `doingWorks `, `doneWorks ` 모두 `WorkViewModel`의 프로퍼티로 처음 초기화가 될 때 어떤 순서로 로드될지 모르기 때문에 프로퍼티에 값을 할당할 때 프로퍼티를 사용할 수 없습니다.

띠리서 사용될 때 값이 생성되는 `lazy` 키워드를 사용하였는데, 현재는 저 프로퍼티 자체를 제거했습니다!

wonhee009 commented on ProjectManager/ProjectManager/View/WorkTableViewCell.swift:
```diff
@@ -0,0 +1,107 @@
+                }
+            })
+            .disposed(by: disposeBag)
+    }
```
`works`는 화면 configure하기 위한 모델 정도로 이해했는데요.
Subject로 정의한 이유가 있을까요?
저는 configure의 역할만 있어 rx가 필요한지에 대해 궁금합니다.

wonhee009 replied:
`WorkManageView`도 동일하게 궁금합니다:)

Judy-999 replied:
RxSwift를 사용하면서 UI 할당 및 변경하는 작업은 모두 rx로 해야 한다고 생각한 것 같습니다 😅

뷰가 굳이 스트림을 가질 필요가 없어 `WorkManageView`와 `WorkTableViewCell` 모두 `configure(with work: Work)` 라는 메서드를 사용하여 뷰를 설정하도록 했습니다!

d354fcb42fffb5aa629b2ab55164347bd451ca5c

wonhee009 commented on ProjectManager/ProjectManager/View/ProjectManagerViewController.swift:
```diff
@@ -0,0 +1,119 @@
+        viewModel.doneCountImage
+            .asDriver(onErrorJustReturn: nil)
+            .drive(projectManagerView.doneTitleView.countImageView.rx.image)
+            .disposed(by: disposeBag)
```
viewModel의 workList에 따라 countImageView 지정을 여기서 해줘도 되지 않나요?
```swift
viewModel.workList.bind {
// 헤더 뷰 이미지 지정
}
```
혹은 viewModel 내 computedProperty로 아이템의 갯수만 가져와도 될거 같구요.

Judy-999 replied:
리스트 별로 `doneCountImage`을 사용하지 않고 `doneWorks`의 count를 이용해 뷰컨트롤러에서 이미지를 지정하도록 변경했습니다!
<br>

```swift
// ProjectManagerViewController.swift
viewModel.doneWorks
    .map {
        UIImage(systemName: "\($0.count).circle.fill")
    }
    .asDriver(onErrorJustReturn: nil)
    .drive(doneTitleView.countImageView.rx.image)
    .disposed(by: disposeBag)
```

721049ee380a0339cd3de9f610b0df9336a9d711

wonhee009 reviewed (changes requested):
수고하셨습니다 🎉 
코멘트 남겨두었으니 확인 부탁드려요~

1. View-ViewController 사이 접근제어자
`projectManageView`에 configure하는 메서드들을 추가하면 될거 같은데요.
```swift
class ProjectManageView {
    private let toDoTitleView = HeaderView()

    func configureToDoTitle(_) { ... }
}
```
이렇게 사용하면 우려되는 부분이 따로 있으실까요?

2. Bind의 방식 및 위치
남겨둔 코멘트가 어느정도 답변이 됐을거 같긴한데요.
뷰는 정말 화면을 그리기만하는 역할이라고 저는 생각합니다.
뷰가 subject 등과 같은 스트림을 가져야할 일이 있을까요?
VC에서 VM의 스트림의 변화를 감지하고 뷰가 화면을 그리기 쉽게 모델을 가공해주는 역할로 생각하고 있습니다.

3. 중복되는 코드 간소화
이런 부분들은 어쩔 수 없다고 생각합니다.
코드 간소화는 할 순 있어도 어떨때는 이렇게 다 적어주는게 가독성이나 코드의 이해도, 코드의 역할을 잘 보여줄 수 있다고 생각합니다.

Judy-999 committed:
chore: 뷰의 초기화 메서드 순서 변경

Judy-999 committed:
refactor: DateManager로 checkOverdue 메서드 분리

Judy-999 committed:
refactor:  ProjectManagerView 제거 및 데이블뷰 헤더 사용

- ProjectManagerViewController에서 뷰를 생성
- WorkTableViewCell을 사용하는 WorkTableView 클래스 생성하여 반복되는 코드 제거
- HeaderView를 테이블뷰의 tableHeaderView로 지정

Judy-999 committed:
refactor: 뷰에서 사용되던 Subject 제거

- 스트림을 만들지 않고 configure 메서드로 뷰의 내용을 채우도록 변경

Judy-999 committed:
refactor: 셀 선택되는 효과 변경

- 회색으로 선택되었다가 바로 해제되도록 변경

Judy-999 committed:
refactor: 헤더의 이미지를 뷰컨트롤러에서 변환해서 넣으주도록 변경

- viewModel에서 CountImage 제거

Judy-999 commented:
라자냐 꼼꼼한 리뷰 감사합니다 😊

rx를 적용하는 것에 급급해 놓쳤던 것이 많았네요 😅

리팩토링 중 커스텀한 뷰를 테이블뷰에 헤더로 넣었는데 아래와 같이 스크롤할 때 같이 스크롤되는 문제가 있습니다.

![헤더 고정문제](https://user-images.githubusercontent.com/102353787/190135527-ecc11a19-99cf-42b3-bb52-46762fe58846.gif)

의미상 제목은 움직이지 않아야 한다고 생각하여 찾아보니 테이블뷰의 style을 `grouped`로 적용하면 헤더가 고정된다고 하는데 어째서인지 해결되지 않습니다 🥲

혹시 헤더를 고정할 수 있는 다른 방법이 있을까요?

wonhee009 commented:
수고하셨습니다~
헤더는 section Header로 넣어주면 위에 달라붙는거로 기억하는데요.
"swift tableView header sticky"로 찾아보시면 힌트를 얻으실 수 있을거 같습니다:)
이 부분은 다음 스텝에서 같이 고민해 봐요!
